_posts/2019-09-25-scatter-gather-pt1.md

```diff
@@ -0,0 +1,94 @@
+        });
+    }
+}
+```
+
+NServiceBus takes care of batching the outgoing messages already. If the handler fails with an exception at any point, no message is sent*.
+
+### Gather:
+Let's say we expect a response message from every previously sent message and we want to continue the process once all responses have been received. This means we need some persistent state to keep track of the expected responses. Very often, people decide to use a [Saga](https://docs.particular.net/nservicebus/sagas) for this, as it provides a shared state across multiple message handlers. The Scatter-Gather implementation would probably look something like this:
+
+```
+class ScatterGather : Saga<ScatterGatherSagaData>,
+    IAmStartedByMessages<OrderPlaced>,
+    IHandleMessages<ProcessOrderItemResult>
+{
+    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ScatterGatherSagaData> mapper)
+    {
+        mapper.ConfigureMapping<OrderPlaced>(m => m.OrderId).ToSaga(s => s.OrderId);
+        mapper.ConfigureMapping<ProcessOrderItemResult>(m => m.OrderId).ToSaga(s => s.OrderId);
+    }
+
+    public async Task Handle(OrderPlaced message, IMessageHandlerContext context)
+    {
+        Data.NumberOfItems = message.OrderItems.Count();
+        foreach (var orderItem in message.OrderItems)
+        {
+            await context.Send(new ProcessOrderItem
+            {
+                OrderId = message.OrderId,
+                ItemId = orderItem.Id
+            });
+        }
+    }
+
+    public async Task Handle(ProcessOrderItemResult message, IMessageHandlerContext context)
+    {
+        var totalProcessedOrders = ++Data.ProcessedOrders;
+        if (totalProcessedOrders == Data.NumberOfItems)
+        {
+            await context.Publish(new OrderProcessed { OrderId = Data.OrderId });
+            MarkAsComplete();
+            Console.WriteLine("Order completed");
+        }
+    }
+}
+```
+
+This looks quite nice, the Saga API provides a simple solution to implement the Scatter-Gather pattern. Unfortunately, it's a very suboptimal solution and you will run into problems. 
+
+### Sagas and concurrency
+
+Sagas are specifically designed to implement long-running business processes, while Scatter-Gather typically isn't very long-running but mostly just a short-lived aggregation process. This short-lived nature is also exactly what causes the problem with the shown implementation. In most cases, scattering is used to parallelize work which means that we expect a lot of responses to arrive back to the saga at the same time. When a handler of a saga is invoked, the data of the saga is loaded from the database. This leads to a situation where multiple responses want to update the same saga instance concurrently (keep in mind that NServiceBus can handle multiple messages concurrently by default). Sagas can't handle concurrency, there is only one message which can successfully complete a saga at one time because they all touch the same data.
```

Comment:
I’m not sure I agree with the definition of long-running. I’d define long running as “spans multiple physical transactions” and not “based on the time it takes”.

```suggestion
In most cases, scattering is used to parallelize work which means that we expect a lot of responses to arrive back to the saga at the same time. When a handler of a saga is invoked, the data of the saga is loaded from the database. This leads to a situation where multiple responses want to update the same saga instance concurrently (keep in mind that NServiceBus can handle multiple messages concurrently by default). Sagas handle concurrency by making sure only one message at a time can successfully update the same data.
```